de.fu_berlin.inf.dpp.server/src/de/fu_berlin/inf/dpp/server/session/NegotiationHandler.java

```diff
@@ -0,0 +1,146 @@
+    private static final Logger log = Logger
+        .getLogger(NegotiationHandler.class);
+
+    private final ISarosSessionManager sessionManager;
+    private final ThreadPoolExecutor sessionExecutor = new ThreadPoolExecutor(0, 4, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerSessionNegotiation-"));
+    private final ThreadPoolExecutor projectExecutor = new ThreadPoolExecutor(0, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(), new NamedThreadFactory(
+        "ServerProjectNegotiation-"));
+
+
+    public NegotiationHandler(ISarosSessionManager sessionManager) {
+        sessionManager.setNegotiationHandler(this);
+        this.sessionManager = sessionManager;
+    }
+
+    @Override
+    public void handleOutgoingSessionNegotiation(
+        final OutgoingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                SessionNegotiation.Status status;
+                if (ServerConfig.isInteractive()) {
+                    status = negotiation
+                        .start(new ConsoleProgressIndicator(System.out));
+                } else {
+                    status = negotiation
+                        .start(new NullProgressMonitor());
+                }
+
+                switch (status) {
+                    case OK:
+                        sessionManager.startSharingProjects(negotiation.getPeer());
+                        break;
+                    case ERROR:
+                        log.error("ERROR running session negotiation: "+negotiation.getErrorMessage());
+                        break;
+                    case REMOTE_ERROR:
+                        log.error("REMOTE_ERROR running session negotiation: "+
+                            negotiation.getErrorMessage()+" at remote: "+negotiation.getPeer().toString());
+                        break;
+                    case CANCEL:
+                        log.info("Session negotiation was cancelled locally");
+                        break;
+                    case REMOTE_CANCEL:
+                        log.info("Session negotiation was cancelled by remote: "+
+                            negotiation.getPeer().toString());
+                        break;
+                }
+            }
+        });
+    }
+
+    @Override
+    public void handleIncomingSessionNegotiation(
+        final IncomingSessionNegotiation negotiation) {
+
+        sessionExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                negotiation.localCancel("Server does not accept session invites",
+                    NegotiationTools.CancelOption.NOTIFY_PEER);
+                negotiation.accept(new NullProgressMonitor());
+            }
+        });
+    }
+
+    @Override
+    public void handleOutgoingProjectNegotiation(
+        final AbstractOutgoingProjectNegotiation negotiation) {
+
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete cases by adding the missing enum constants or add a default case to this switch. ([squid:SwitchLastCaseIsDefaultCheck](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=144&issues=AWbKFPs6O08i52vdg2B1&open=AWbKFPs6O08i52vdg2B1)

[ISSUE_KEY]: AWbKFPs6O08i52vdg2B1